Sources/Build/ToolProtocol.swift

```diff
@@ -61,22 +61,22 @@ struct SwiftcTool: ToolProtocol {
         // For C family targets Swift needs dynamic libraries to be able to interpolate.
         // We implicitly create dynamic libs for all C targets ie ClangModules, add
         // input to the product and not the module for ClangModules.
-        return module.recursiveDependencies.map{ module in
+        return module.sources.paths.map{ $0.asString } + module.recursiveDependencies.flatMap{ module in
             switch module {
-            case let module as SwiftModule:
-                return module.targetName
             case let module as ClangModule:
                 let product = Product(name: module.name, type: .Library(.Dynamic), modules: [module])
-                return product.targetName
-            case let module as CModule:
-                return module.targetName
+                return prefix.appending(product.outname).asString
+            case let module as SwiftModule:
+                return prefix.appending(component: module.c99name + ".swiftmodule").asString
```

Comment:
Agreed.  I'll factor it out into a module method.  This entire logic needs reworking so that paths are specified only once.  I'm working on a rewrite but it is not finished yet, and meanwhile there seemed to be benefit to fixing the virtual node issue even in the current code.